File: AlgorithmsPackCPP/Restore the Array From Adjacent Pairs.cpp
class Solution {
    typedef unordered_map<int,vector<int>>GRAPH;

    void addEdge(GRAPH&graph, int from, int to){
        graph[from].push_back(to);
        graph[to].push_back(from);
    }
public:
    void dfs(GRAPH&graph, int node, unordered_set<int>&visited, vector<int>&ccNodes){
        ccNodes.push_back(node);
        visited.insert(node);

        for(int cur:graph[node]){
            if(!visited.count(cur)){
                dfs(graph,cur,visited,ccNodes);
            }
        }
    }

    vector<int> restoreArray(vector<vector<int>>& adjacentPairs) {
        GRAPH graph;
        for(auto&a:adjacentPairs)
            addEdge(graph,a[0],a[1]);

        unordered_set<int>visited;
        vector<int>ccNodes;

        for(auto&node:graph){
            if(!visited.count(node.first)){
                dfs(graph,node.first,visited,ccNodes);
                break;
            }
        }
        return ccNodes;
    }
};

File: AlgorithmsPackJavaSolutions/Restore the Array From Adjacent Pairs.java
class Solution {
    private void addEdge(Map<Integer, List<Integer>> graph, int from, int to) {
        graph.computeIfAbsent(from, key -> new ArrayList<>()).add(to);
        graph.computeIfAbsent(to, key -> new ArrayList<>()).add(from);
    }

    private void dfs(Map<Integer, List<Integer>> graph, int node, Set<Integer> visited, List<Integer> result) {
        visited.add(node);
        result.add(node);

        for (int neighbour : graph.get(node)) {
            if (!visited.contains(neighbour)) {
                dfs(graph, neighbour, visited, result);
            }
        }
    }

    public int[] restoreArray(int[][] adjacentPairs) {
        Map<Integer, List<Integer>> graph = new HashMap<>();

        for (int[] pair : adjacentPairs) {
            addEdge(graph, pair[0], pair[1]);
        }

        Set<Integer> visited = new HashSet<>();
        List<Integer> result = new ArrayList<>();

        for (Map.Entry<Integer, List<Integer>> entry : graph.entrySet()) {
            if (entry.getValue().size() == 1) {
                dfs(graph, entry.getKey(), visited, result);
                break;
            }
        }

        int[] restoredArray = new int[result.size()];
        for (int i = 0; i < result.size(); i++) {
            restoredArray[i] = result.get(i);
        }

        return restoredArray;
    }
}
